libsolidity/codegen/ExpressionCompiler.cpp

```diff
@@ -1289,14 +1292,22 @@ bool ExpressionCompiler::visit(MemberAccess const& _memberAccess)
 			_memberAccess.expression().accept(*this);
 		return false;
 	}
-	// Another special case for `this.f.selector` which does not need the address.
-	// There are other uses of `.selector` which do need the address, but we want this
-	// specific use to be a pure expression.
+	// Another special case for `this.f.selector` and for ``C.f.selector`` which do not need the address.
+	// There are other uses of `.selector` which do need the address, but we want these
+	// specific uses to be pure expressions.
 	if (
-		_memberAccess.expression().annotation().type->category() == Type::Category::Function &&
-		member == "selector"
+		auto const* functionType = dynamic_cast<FunctionType const*>(_memberAccess.expression().annotation().type);
```

Comment:
I'm still not used to this. Maybe we could change the style such that the `;` has to be on a line of its own in these cases?

Reply:
Yeah, I hesitated a bit with this :-). Mainly did it here to avoid blowing up the diff with indentation-only changes below... which is probably not the best reason...

In general I like it, though - if it were up to me alone, I'd e.g. also merge the one a few lines down to
```
			if (
				auto const* exprInt = dynamic_cast<Identifier const*>(&expr->expression()));
				exprInt && exprInt->name() == "this"
			)
```

Less nesting always seems nicer to me... And less nesting in ``if``s is likely to avoid brackets arising from otherwise ambiguous ``else``s... but yeah, I do see the point about potentially forgetting the explicit null check, so not sure...

Not sure I'd still like it, if I had to use a whole line for the ``;``, though :-)... Would that really help other than it probably being looked at twice, because it looks weirder :-)?

I can nest here for now, if you prefer and until we decided how to deal with this option in general...

Reply:
It's fine! I would just like to have this very different kind of if statement stand out a bit more. The fact that you have to read it differently is only apparent towards the end. Would have preferred something like

```
if <statement> (condition) { ... }
```